kernel/src/syscall.rs

```diff
@@ -1,21 +1,108 @@
-//! Tock syscall number definitions.
+//! Tock syscall number definitions and arch-agnostic interface trait.
+
+use process;
 
 /// The syscall number assignments.
 #[derive(Copy, Clone, Debug)]
-crate enum Syscall {
+pub enum Syscall {
     /// Return to the kernel to allow other processes to execute or to wait for
     /// interrupts and callbacks.
-    YIELD = 0,
+    ///
+    /// SVC_NUM = 0
+    YIELD,
 
     /// Pass a callback function to the kernel.
-    SUBSCRIBE = 1,
+    ///
+    /// SVC_NUM = 1
+    SUBSCRIBE {
+        driver_number: usize,
+        subdriver_number: usize,
+        callback_ptr: *mut (),
+        appdata: usize,
+    },
 
     /// Instruct the kernel or a capsule to perform an operation.
-    COMMAND = 2,
+    ///
+    /// SVC_NUM = 2
+    COMMAND {
+        driver_number: usize,
+        subdriver_number: usize,
+        arg0: usize,
+        arg1: usize,
+    },
 
     /// Share a memory buffer with the kernel.
-    ALLOW = 3,
+    ///
+    /// SVC_NUM = 3
+    ALLOW {
+        driver_number: usize,
+        subdriver_number: usize,
+        allow_address: *mut u8,
+        allow_size: usize,
+    },
 
     /// Various memory operations.
-    MEMOP = 4,
+    ///
+    /// SVC_NUM = 4
+    MEMOP { operand: usize, arg0: usize },
+}
+
+/// Why the process stopped executing and execution returned to the kernel.
+pub enum ContextSwitchReason {
+    /// Process exceeded its timeslice, otherwise catch-all.
+    Other,
```

Comment:
Is `Other` ever used for reasons apart from timeslices expiring? It seems like we should name it for its purpose `TimesliceExpired` (or similar). If there really are other possible reasons, we should have a catch-all for them apart from timeslices.

Reply:
I don't know.

Reply:
+1 for no `Other`

Reply:
I agree with replacing `Other` with something timeslice specific.

From an API design perspective, I think it doesn't cost anything to _add_ specific reasons later, since the implementer of `UserspaceKernelBoundary` only _produces_ values of this type, and doesn't pattern match. So if some architecture might have some other reason to interrupt the process at some point later, we can add that reason to the enum with no changes needed to previous implementations.

Reply:
How does a timeslice expiration manifest? Put another way, does the implementer of `get_and_reset_context_switch_reason()` know that the reason the process stopped was because of the timeslice expiration? This is not clear in the original implementation, and why `Other` was created.

Reply:
You're right...

the systick timer fires, which is also in the arch crate. That logic, as you say, is basically split now between the arch crate (handling the systick interrupt and setting a global flag shared with the kernel crate), and the kernel crate reading the global flag and deciding---yep, this process exceeded it's timeslice.

Reply:
Is there a global flag? Or is it the absence of a global flag being set that indicates it was the systick firing?